### ISSUE TITLE
Request for different X-Forwarded-For modes 

### ISSUE THREAD

**Issue by mback2k** [traefik/traefik#6095]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Besides the existing issues https://github.com/containous/traefik/issues/4099 (new header standard) and https://github.com/containous/traefik/issues/5482 (possibility to remove the header) I would like to see the possibility to modify the header, either via regex or by choosing between different modes.

The reason for that is that there are applications that do interpret `X-Forwarded-For` in various ways. Some follow the RFCs and interpret the list of IPs from left-to-right as client-to-server, some do it the other way around and others only trust the right-most or left-most value. Therefore I would like to be able to make traefik drop existing values in order to have only one IP address value or change the order of the IP address values. I know that it should be the applications that adopt the standards correctly, but sometimes that is not so easy with legacy or closed source applications. One example I would like to see and hope to have fixed is [Home Assistant](https://github.com/home-assistant/home-assistant/issues/29868).

To summarise again, I would imagine setting the following forwardedHeaders modes:

- left-to-right (default)
- right-to-left (to change the order of IP addresses)
- drop-existing (to drop incoming header IP address[es])
- use-existing (to keep incoming header in place and IP address[es] untouched)

**Comment by ldez** [comment]
Hello,

I think you are looking for https://docs.traefik.io/v2.1/routing/entrypoints/#forwarded-header

**Comment by mback2k** [comment]
Unfortunately that is not providing all the modes mentioned above. Yes, it is possible to discard incoming X-Forwarded-For (drop-existing) by not trusting it, but changing the order or just using the existing value without adding to it is not possible (yet).

**Comment by Shagon94** [comment]
@mback2k to clarify, it is currently not possible to have traefik only proxy headers that were part of the original request instead of automatically adding it?

Example I see that traefik by default is adding:

```yml
X-Forwarded-For
X-Forwarded-Host
X-Forwarded-Port
X-Forwarded-Proto
X-Forwarded-Server
X-Real-Ip
```

Would it be possible to turn off those and only proxy them when they were part of the original request, so:

```json
$ curl localhost:40001/api/v1/echo -s | jq
{
  "Accept": "*/*",
  "Host": "localhost:40001",
  "User-Agent": "curl/7.47.0"
}
```
By default curl uses only those 3 values in the request, if I were to do the same via traefik:

```js
{
  "Accept": "*/*",
  "Accept-Encoding": "gzip",
  "Host": "host.com",
  "User-Agent": "curl/7.47.0",
  "X-Forwarded-For": "1.2.3.4", // (my real ip)
  "X-Forwarded-Host": "host.com",
  "X-Forwarded-Port": "80",
  "X-Forwarded-Proto": "http",
  "X-Forwarded-Server": "myhost",
  "X-Real-Ip": "1.2.3.4"  // (my real ip)
}
```

The thing I had in mind was to only proxy if supplied as so:

```js
$ curl localhost:40001/api/v1/echo -s -H "X-forwarded-for: 1.2.3.4" | jq
{
  "Accept": "*/*",
  "Host": "localhost:40001",
  "User-Agent": "curl/7.47.0",
  "X-Forwarded-For": "1.2.3.4"
}
```

**Comment by mback2k** [comment]
@Shagon94 I think you are talking about the use-case "use-existing (to keep incoming header in place and IP address[es] untouched)" mentioned in my first comment, but I am not sure what the output in your 3rd code box ("if I were to do the same via traefik:") represents?

So, yes, 1) I basically would like to tell traefik to just keep and forward the existing headers by not changing them. And on top of that 2) I would also like to see an option to tell it to append it's values on the left instead of the right, because there are unfortunately applications out there that interpret the proxy headers in the wrong order.

**Comment by mwitkow** [comment]
I have a similar case, that would benefit from this use case:

I'd like Traefik to be my auth proxy in front of Home Assistant, which only really allows for customized sign-ins based on IP addresses (https://www.home-assistant.io/integrations/http/).

I'd really like to be able to overwrite the full `X-Forwarded-For` that the downstream application sees. Using custom headers, I can only left-prepend an `X-Forwarded-For` using custom headers.

The downstream application seees the following:
```
GET / HTTP/1.1
Host: guest.example.com
User-Agent: curl/7.68.0
Accept: */*
X-Forwarded-For: 127.0.100.2, 192.168.1.1
...
Accept-Encoding: gzip
```
And unfortunately interprets the `192.168.1.1` as the client ip for auth purposes rather than the manually added `192.168.1.1`. Would be great to be able to control the whole thing the downstream application sees.

**Comment by CybotTM** [comment]
Hi @mwitkow 

> And unfortunately interprets the `192.168.1.1` as the client ip for auth purposes rather than the manually added `127.0.100.2`. Would be great to be able to control the whole thing the downstream application sees.

You need to add `192.168.1.1` to your trusted proxy list. This should suffice, if the app follows RFC to determine client IP.

**Comment by EmilioMoreno** [comment]
I would like to +1 to either have Traefik not add its own ip to the existing X-Forwarded-For header (use-existing), or actually be able either regex modify the header.

In my case, Traefik its adding itself as 127.0.0.1,  which makes no sense when the header is processed at the endpoint, so I rather prefer either Traefik not modifying X-Forwarded-For at all, or be able to manipulate it to remove or specify the right ip

While if have been searching also some plugins to modify an existing request header, none of them work or are completely outdated.

**Comment by kevinpollet** [comment]
Hello @EmilioMoreno,

> I would like to +1 to either have Traefik not add its own ip to the existing X-Forwarded-For header (use-existing), or actually be able to regex modify the header.

Traefik is following the de-facto standard explained in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For, and is only adding the client IP to the `X-Forwarded-For` header. For more information on trusting forwarded headers, see https://doc.traefik.io/traefik/routing/entrypoints/#forwarded-headers

**Comment by kevinpollet** [comment]
Hello all,

After discussing this proposal with the team, we think that Traefik should only follow the de-facto standard explained in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For, and allowing to change that is not something we want as it would be confusing.
